Fussing with unit test encapsulation [Experimental]

agilliland opened the pull request:
Primary goal here is 2 things:
1. completely purge all app data between each unit test execution to ensure that tests are fully encapsulated and don't affect each other.
2. provide a simple and lightweight way to define a test which specifies it's own arbitrary set of data to use with the test execution.

The inspiration for this is from a variety of places such as:

```
(do
    ;; Delete all the other random Users we've created so far
    (let [user-ids (set (map user->id [:crowberto :rasta :lucky :trashbird]))]
      (cascade-delete User :id [not-in user-ids]))
    ;; Now do the request
    (set ((user->client :crowberto) :get 200 "user")))
```

:cry: we have a lot of code in unit tests which is there purely to cleanup existing data and attempt to create a concrete dataset for its own test.  this is both ugly and error prone.

```
(when-let [user (create-user-api rand-name)]
      ;; create a random user then set them to :inactive
      (upd User (:id user)
        :is_active false
        :is_superuser true)
      ;; then try creating the same user again
      ((user->client :crowberto) :post 200 "user" {:first_name (:first_name user)
                                                   :last_name "whatever"
                                                   :email (:email user)}))
```

:disappointed: we repeat ourselves quite a bit from test to test specifically to create simple things like test objects and that muddies the reading of the unit tests and duplicates code unnecessarily.

:smile: instead of both of these scenarios we can add in a little code the resets the app db for each unit test and then we can provide a simple macro which allows us to create our desired test data with the test execution.  for example:

```
(def single-example-user 
    {:before (fn []
                    (ins User :first_name "Rasta" :last_name "Toucan" :email "rasta@metabase.com"))})

(expect-with-test-data single-example-user
    "Rasta"
    (sel :one :field [User :first_name] :email "rasta@metabase.com"))

(expect-with-test-data single-example-user
    "Toucan"
    (sel :one :field [User :last_name] :email "rasta@metabase.com"))
```

agilliland committed:
very basic function for clearing out MB database before each unit test starts.

agilliland committed:
fix up test code for automatically creating users on demand.

agilliland committed:
add some basic macros for `expect-with-test-db` and `expect-with-test-data` which allow us to more easily define unit tests which do some prep work for setting up data the test needs to run.

agilliland committed:
simplify and streamline a couple things.

agilliland committed:
example of using `expect-with-test-db` on sync tests.

agilliland committed:
quick example of using `expect-with-test-data` for a unit test in user api test.

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -212,3 +213,48 @@
+    `(def ~(vary-meta fn-name assoc :expectation true)
+       (fn []
+         (with-temp-db ~test-db-binding
+           (expectations/doexpect ~expected ~actual))))))
```
You could simplify this macro a lot by just using `defn`:

```clojure
(defmacro expect-with-test-db [test-db-binding expected actual]
  `(defn fn# {:expectation true} []
     (with-temp-db ~test-db-binding
       (expectations/doexpect ~expected ~actual))))
```
*  `def + fn = defn`, you could just use that directly here
*  `defn` takes an optional metadata map param inbetween the (optional) docstr and the args. So no need to `vary-meta` 
*  Since you don't need to `vary-meta` you can just do an inline auto gensym

:+1:

camsaul commented on test/metabase/test_setup.clj:
```diff
@@ -75,6 +76,37 @@
+                         "metabase_database"
+                         "core_user"]
+            mk-del-stmt (partial str "delete from ")]
+        (dorun (map (comp (partial jdbc/db-do-prepared conn) mk-del-stmt) tables))))
```
Don't get carried away with those `comps` and `partials` :wink: . The prevailing wisdom in Clojure-land is to prefer higher-level abstractions like `for` or `doseq` where they make sense. You could just write:
```
(doseq [table tables]
  (jdbc/execute! conn [(str "delete from " table)]))
```

A bit easier to wrap your head around :scream_cat: :+1:

camsaul replied:
Also this might just be me but I find the same form upwards of 3x as easy to read if you stick a newline in there.
Then you have both args to `map` aligned and it's really easy to set what's going on. 

```clojure
(dorun (map (comp (partial jdbc/db-do-prepared conn) mk-del-stmt)
            tables))
```

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -212,3 +213,48 @@
+    (finally
+      (binding [*sel-disable-logging* true]
+        (when (fn? (:after i-test-data))
+          ((:after i-test-data)))))))
```
I would write this as 
```clojure
(defn with-test-data [{:keys [before after]} i-test-data f]
  (try
    (when (fn? before)
      (binding [*sel-disable-logging* true]
        (before)))
    (f)
    (finally
      (when (fn? after)
        (binding [*sel-disable-logging* true]
          (after))))))
```

*  Setting up bindings are a bit expensive what with all the thread-safety and try-finallys going on. It looks like in the majority of cases these bindings aren't going to be used at all :disappointed: . Seeing how this code could easily go either way, I would vote on nesting them inside the `whens`. This way they're only being established in cases where they're being used
*  If you think about it, it doesn't make sense to establish a binding around a pair of forms, and then establish a binding in the second to reverse the original. :unamused: 
*  I think this code is upwards of 50% more readable if you destructure `i-test-data` so you can just access `before` and `after` directly :flushed:

camsaul replied:
Wait a second. Here's an even better way to write it:

```clojure
(defn with-test-data [{:keys [before after]} f]
  (let [no-log (fn [x] (binding [*sel-disable-logging* true] (x)))]
    ((as-> f f
       (if-not (fn? before) f (fn [] (no-log before) (f)))
       (if-not (fn? after)  f (fn [] (try (f)
                                           (finally (no-log after)))))))))
```

This way you can avoid wrapping the whole thing in a try-catch if `after` isn't being used

agilliland commented:
Cool.  So apart from the much needed improvements to my macro writing skills, the bigger question is do we like this approach to structuring the unit tests?

If we are going to commit to something like this and basically have to go back through quite a few of our unit tests and update them then I want to make sure we think this really is the best approach.

camsaul commented:
This is definitely a better approach but I want to know how it will affect:

1. *Ease of writing tests*: It's very important this doesn't make our tests any harder to write. It would be an international travesty if we left each test to fend for itself with regards to loading data or otherwise getting Metabase into a useful state. I don't want to go back to the old Django-tard days when we'd have 100+ lines of copypasta in every test class just to setup a DB with some  users just to run basic tests. Those tests took so long to write we just didn't do it half the time.

2.  *Performance*: We have nearly a thousand unit tests when you run against all the drivers, so we should be careful when adding code to run around each one. Circle is slower than dirt so if it adds a minute on your local box you can probably assume it's going to be 3 or 4 on Circle.

agilliland commented:
Yeah, I definitely want to make sure this is improving things, which is why I bring it up.

Re: Easy of writing tests.  This is priority #1 IMO and should be the main design decision.  I want the unit tests to be concise and easy to write so that we are encouraged to write & maintain them.  This is my biggest hang-up right now is that I don't particularly feel like it's particularly easy to write tests in our current codebase and I'd like to improve that.

Beyond ease of writing the tests I'd say that consistency of execution is my #2 priority.  My biggest overall issue with our current unit tests is that from test to test they all share the same db and the data is never cleared between tests, so effectively any test we write can impact another test and that's a problem.  It also leads to writing ugly tests because many of our tests have code that is only there to try and clear out old data from other tests and create a sane environment to run in.  This is the main reason why I tried the approach of deleting all of the data in the db as part of each test.

Re: Performance.  I personally consider this a little lower priority.  Obviously we can't have tests that take 30 minutes, but if we get a bunch of other improvements for 1-2 minutes longer test execution then I can live with that.  Besides, I think we can always address this through parallelizing more of the testing.

Getting back to the code, my hope is that if we have a nice reusable structure for defining a test which has a set of test data that it wants loaded at start then that should be pretty concise and highly reusable.

e.g.

```
(def user-test-data {:before (fn []
                               (ins User
                                 :first_name "abc"
                                 :last_name "def"
                                 :email "abc@metabase.com"
                                 :password "def"
                                 :is_superuser false))})

(expect-with-test-data user-test-data
  "abc"
  (sel :one :field [User :first_name]))

(expect-with-test-data user-test-data
  "def"
  (sel :one :field [User :last_name]))

(expect-with-test-data user-test-data
  "abc@metabase.com"
  (sel :one :field [User :email]))
```

So the idea would be that you define 1 or 2 test datasets at the top of a testing namespace and then reuse that throughout the tests you write.  The example above is contrived and not super useful, but if the test data included more objects such as 2-3 users with different options then you could easily see a case where you could write a dozen unit tests off that same data.

agilliland commented:
BTW, here's a perfect example of a frustration in action.  I just added a new function in our setting code for `set-all` which allows the caller to easily set multiple settings at the same time.  After writing this I added a unit test for it which is working fine, but then I find out that it broke another unit test ...

```
failure in (setting_test.clj:131) : metabase.models.setting-test
(expect
 {:test-setting-2 "TOUCANS"}
 (do
  (set-settings nil "TOUCANS")
  (m/filter-keys
   (fn* [p1__31724#] (re-find #"^test-setting-\d$" (name p1__31724#)))
   (setting/all))))

           
expected:
 {:test-setting-2 "TOUCANS"}
 
was:
 {:test-setting-3 "I win!",
 :test-setting-4 "For realz",
 :test-setting-2 "TOUCANS"}


           
in actual, not expected:
{:test-setting-3 "I win!", :test-setting-4 "For realz"}
```

This is why I think it's important that we clear the data from the db on each unit test.  It's really frustrating to have to worry about disrupting other unit tests in the system every time you create a new test.

agilliland commented:
closing this since it's super old
